doc/design/diagrams/request_response/typed_api.puml

```diff
@@ -0,0 +1,98 @@
+@startuml
+skinparam backgroundColor #EEEBDC
+
+together Group1 {
+class Client<Req, Res> {
+  Client(serviceDescription: ServiceDescription&, clientOptions: ClientOptions&)
+  connect(): void
+  disconnect(): void
+  getConnectionState(): ConnectionState
+  loan(): expected<Request<Req>, Error>
+  send(request: Request<Req>): void
+  hasResponses(): bool
+  take(): expected<Response<Res>, Error>
+  releaseQueuedResponses(): void
+  hasMissedResponses(): bool
+}
+
+class Server<Req, Res> {
+  Server(serviceDescription: ServiceDescription&, serverOptions: ServerOptions&)
+  offer(): void
+  stopOffer(): void
+  isOffered(): bool
+  hasClients(): bool
+  hasRequests(): bool
+  take(): expected<Request<Req>, Error>
+  releaseQueuedRequests(): void
+  hasMissedRequests(): bool
+  loan(request: Request<Req>&): expected<Response<Res>, Error>
+  send(response: Response<Res>): void
+}
+
+class Publisher {
+}
+}
+
+class SmartChunk<TransmissionInterface, T, H> {
+  SmartChunk(chunkUniquePtr: unique_ptr<T>&&, sender: TransmissionInterface&)
+  SmartChunk(chunkUniquePtr: unique_ptr<const T>&&)
+  operator->(): T*
```

Comment:
We had once discussed that every container/data type which may contain something or contains either the one value or the other uses additionally the `and_then`/`or_else` pattern. Data types like `optional`, `expected`, `unique_ptr` and of course our `SmartChunk` here.

I would extend this interface with the following methods:
```cpp
SmartChunk& and_then(const cxx::function_ref<void(T&)> & callback) noexcept;
SmartChunk& and_then(const cxx::function_ref<void(const T&)> & callback) const noexcept;
SmartChunk& or_else(const cxx::function_ref<void()> & callback) const noexcept;
```

Reply:
I would decline this request. If you have a null-able type use `cxx::optional<T>` instead. We should not reinvent null-able types all over the place since that essentially means you always have to do a null check. If it is wrapped in an optional you clearly state your intention.

Reply:
Oh, and I'm aware that due to to C++ move semantic we are left with undefined but specified objects. But this objects are usually confided in a local scope where it is easy to reason about the movement and it can also easily be tested that moved objects are not used after the move operation.